Installing the Dapr Helm Chart via the CLI

superbeeny opened the pull request:
# Description

Installing the Dapr Helm Chart via the CLI

## Type of change

- This pull request adds a feature to install Dapr to its own namespace by using the rad cli. 

Fixes: #8032 

## Contributor checklist
Please verify that the PR meets the following requirements, where applicable:

- [X] An overview of proposed schema changes is included in a linked GitHub issue.
- [ ] A design document PR is created in the [design-notes repository](https://github.com/radius-project/design-notes/), if new APIs are being introduced.
- [ ] If applicable, design document has been reviewed and approved by Radius maintainers/approvers.
- [ N/R ] A PR for the [samples repository](https://github.com/radius-project/samples) is created, if existing samples are affected by the changes in this PR.
- [ N/R ] A PR for the [documentation repository](https://github.com/radius-project/docs) is created, if the changes in this PR affect the documentation or any user facing updates are made.
- [ N/R ] A PR for the [recipes repository](https://github.com/radius-project/recipes) is created, if existing recipes are affected by the changes in this PR.

superbeeny committed:
WIP: Implementing Dapr install for radius

Signed-off-by: Nick Beenham <Nicholas_Beenham@cable.comcast.com>

superbeeny committed:
working commit

Signed-off-by: Nick Beenham <Nicholas_Beenham@cable.comcast.com>

superbeeny committed:
fixing options initialization

Signed-off-by: Nick Beenham <Nicholas_Beenham@cable.comcast.com>

ytimocin commented on pkg/cli/cmd/radinit/cluster.go:
```diff
@@ -39,11 +39,11 @@ func (r *Runner) enterClusterOptions(options *initOptions) error {
 		return clierrors.MessageWithCause(err, "Unable to verify Radius installation.")
 	}
-	options.Cluster.Install = !state.Installed
+	options.Cluster.Install = !state.RadiusInstalled || !state.DaprInstalled
```
Does this mean that if **Dapr** is not installed we can't run `rad init`?

superbeeny replied:
Do you mean if radius is installed but dapr is not?

ytimocin replied:
Got my answer to my question :) Thanks @superbeeny!

ytimocin commented on bicep-types:
PR needs a rebase

rynowak replied:
Yup. This shouldn't be changing in this PR.

ytimocin commented on pkg/cli/cmd/radinit/cluster.go:
```diff
@@ -39,11 +39,11 @@ func (r *Runner) enterClusterOptions(options *initOptions) error {
 		return clierrors.MessageWithCause(err, "Unable to verify Radius installation.")
 	}
-	options.Cluster.Install = !state.Installed
+	options.Cluster.Install = !state.RadiusInstalled || !state.DaprInstalled
```
Should we add a test for this new case?

ytimocin commented on pkg/cli/cmd/uninstall/kubernetes/kubernetes.go:
```diff
@@ -91,13 +92,21 @@ func (r *Runner) Run(ctx context.Context) error {
+		Radius: helm.ChartOptions{
+			Namespace: helm.RadiusSystemNamespace,
+		},
+		Dapr: helm.ChartOptions{
```
Is this only going to delete Dapr if it is installed via Radius? What if I have a Dapr installation on my cluster that I installed manually? Is that also going to be deleted if I run `rad uninstall kubernetes`?

superbeeny replied:
no it won't be deleted, though my feelings are that we should clean up after ourselves, but it gets complicated during upgrades etc

rynowak replied:
> What if I have a Dapr installation on my cluster that I installed manually?

The problem is exactly this. We won't later whether or not we're the ones that installed Dapr. 

We should do the safe thing.

rynowak replied:
Following up on this discussion, we investigated and it turns out Radius also uninstall Contour when removed. 

Since both decisions have downsides, we're going to be consistent and also uninstall Dapr until we get user feedback to the contrary.

ytimocin commented on pkg/cli/helm/cluster.go:
```diff
@@ -87,49 +98,44 @@ func PopulateDefaultClusterOptions(cliOptions CLIClusterOptions) ClusterOptions
-	namespace := RadiusSystemNamespace
 	flags := genericclioptions.ConfigFlags{
-		Namespace: &namespace,
+		Namespace: &clusterOptions.Radius.Namespace,
```
Nice catch

ytimocin commented on pkg/cli/helm/cluster.go:
```diff
@@ -143,55 +149,90 @@ func UninstallOnCluster(kubeContext string) error {
-// CheckRadiusInstall checks if the Radius release is installed in the given kubeContext and returns an InstallState object
-// with the version of the release if installed, or an error if an error occurs while checking.
-func CheckRadiusInstall(kubeContext string) (InstallState, error) {
+// QueryRelease checks to see if a release is deployed to a namespace for a given kubecontext.
```
nit
```suggestion
// queryRelease checks to see if a release is deployed to a namespace for a given kubecontext.
```

ytimocin replied:
What are the 3 outputs of this function? Can we add a doc for that too?

ytimocin commented on pkg/cli/helm/cluster.go:
```diff
@@ -143,55 +149,90 @@ func UninstallOnCluster(kubeContext string) error {
+// CheckRadiusInstall checks if the Radius release is installed in the given kubeContext and returns an InstallState object
+// with the version of the release if installed, or an error if an error occurs while checking.
+func CheckRadiusInstall(kubeContext string) (InstallState, error) {
+
```
nit: extra line can be removed

rynowak commented on pkg/cli/helm/cluster.go:
```diff
@@ -33,13 +33,15 @@ const (
 	ContourChartDefaultVersion = "11.1.1"
 )
 
+// TODO: Revisit and reevaluate
```
Is this a note for your PR or something else?

rynowak commented on pkg/cli/helm/cluster.go:
```diff
@@ -143,55 +149,90 @@ func UninstallOnCluster(kubeContext string) error {
+	if err != nil {
+		return fmt.Errorf("failed to get helm config, err: %w, helm output: %s", err, helmOutput.String())
+	}
+	err = RunHelmUninstall(daprHelmConf, clusterOptions.Dapr)
```
I thought we weren't uninstalling Dapr?

superbeeny replied:
dapr is not uninstalled

superbeeny committed:
adding default value to options

Signed-off-by: Nick Beenham <Nicholas_Beenham@cable.comcast.com>

superbeeny commented:
After some discussion with @rynowak and a review of the current codebase where contour is uninstalled we agreed to be consistent and uninstall dapr

rynowak commented:
> After some discussion with @rynowak and a review of the current codebase where contour is uninstalled we agreed to be consistent and uninstall dapr

Yes 👍. I was under the impression that we left Contour alone during uninstall. It turns out we remove it. 

@superbeeny and I made the executive decision to be consistent with our existing behavior in the absence of user feedback. Both options (leave-installed, uninstall) have downsides. If we get feedback we'll change this.

I'll remind everyone that `rad uninstall` is provided for convenience, and so we're optimizing for convenience. We expect users to use the Helm charts directly in more serious scenarios.

rynowak reviewed (dismissed):
Changes look good to me 👍 We should merge this now that we've branched for 0.39. We will be layering other changes on top for 0.40.

ytimocin commented:
These [workflows](https://github.com/search?q=repo%3Aradius-project%2Fradius+%22Install+Radius%22+language%3AYAML&type=code&l=YAML) may need to be tested/updated since Dapr is installed via Helm in those.

I see some functional tests of this PR fail at the Install Radius step: https://github.com/radius-project/radius/actions/runs/11865183744/job/33070281563?pr=8033.

superbeeny committed:
fixing the install tests

Signed-off-by: Nick Beenham <Nicholas_Beenham@cable.comcast.com>

superbeeny committed:
fixing last unit test

Signed-off-by: Nick Beenham <Nicholas_Beenham@cable.comcast.com>

superbeeny commented:
> workflows

My first guess is that we're not detecting that dapr is already installed, should be easy enough to replicate ... thanks!!

superbeeny committed:
commenting out dapr install in functional tests

Signed-off-by: Nick Beenham <Nicholas_Beenham@cable.comcast.com>

superbeeny committed:
Merge branch 'main' into features/superbeeny/dapr-install

superbeeny committed:
Fixing #8043 - adding long form commands and aliasing abbreviations (#8073)

# Description

Updated cmd line functions to use `environment` in place of `env` and
`initialize` in place of `init`. Both abbreviations are now aliased so
this is a non breaking change.

## Type of change

- This pull request fixes a bug in Radius and has an approved issue
(issue link required).

Fixes: #8043 

Signed-off-by: Nick Beenham <Nicholas_Beenham@cable.comcast.com>

rynowak commented on .github/workflows/functional-test-cloud.yaml:
```diff
@@ -595,10 +595,10 @@ jobs:
+      # - name: Install dapr into cluster
+      #   run: |
+      #     wget -q https://raw.githubusercontent.com/dapr/cli/master/install/install.sh -O - | /bin/bash -s ${{ env.DAPR_VER }}
+      #     dapr init -k --wait --timeout 600 --runtime-version ${{ env.DAPR_VER }} --dashboard-version ${{ env.DAPR_DASHBOARD_VER }}
```
Hey @superbeeny - I have a question about removing this. 

Does `rad install kubernetes` install fail when Dapr is already installed? I think we should make sure that case succeeds. 

I'm trying to assess whether this is being removed because it's unnecessary, or whether it's being removed because it's **broken**.

superbeeny replied:
Rad init will not fail for an existing dapr install

rynowak reviewed (commented):
One more question here: https://github.com/radius-project/radius/pull/8033/files#r1851230293

I think this is almost ready to go.

superbeeny commented on pkg/cli/helm/cluster.go:
```diff
@@ -87,49 +97,44 @@ func PopulateDefaultClusterOptions(cliOptions CLIClusterOptions) ClusterOptions
 
-	err = ApplyContourHelmChart(options.Contour, kubeContext)
+	// Install Dapr
+	daprFound, err := ApplyHelmChart(clusterOptions.Dapr, kubeContext)
```
done

superbeeny committed:
removing comments and adding version for dapr

Signed-off-by: Nick Beenham <Nicholas_Beenham@cable.comcast.com>

superbeeny committed:
Merge branch 'main' into features/superbeeny/dapr-install

rynowak reviewed (approved):
:shipit: - this is awesome! thanks @superbeeny
